packages/belos/src/BelosSolverFactory.hpp

```diff
@@ -389,161 +450,48 @@ class SolverFactory :
-  }
-  case SOLVER_TYPE_BICGSTAB: {
-    typedef BiCGStabSolMgr<Scalar, MV, OP> impl_type;
-    return makeSolverManagerTmpl<base_type, impl_type> (params);
-  }
-  default: // Fall through; let the code below handle it.
-    TEUCHOS_TEST_FOR_EXCEPTION(
-      true, std::logic_error, "Belos::SolverFactory: Invalid EBelosSolverType "
-      "enum value " << solverType << ".  Please report this bug to the Belos "
-      "developers.");
-  }
+namespace Impl {
 
-  // Compiler guard.  This may result in a warning on some compilers
-  // for an unreachable statement, but it will prevent a warning on
-  // other compilers for a "missing return statement at end of
-  // non-void function."
-  TEUCHOS_UNREACHABLE_RETURN(Teuchos::null);
+// specialize for Belos registration - such as for Belos_Factory unit test
+template<>
+inline void
+SolverFactoryParent<double, MultiVec<double>, Operator<double> >::
+registerSolverFactoryForLib () {
+  Belos::Details::registerSolverFactory();
 }
 
-template<class SolverManagerBaseType, class SolverManagerType>
-Teuchos::RCP<SolverManagerBaseType>
-makeSolverManagerTmpl (const Teuchos::RCP<Teuchos::ParameterList>& params)
-{
-  using Teuchos::ParameterList;
-  using Teuchos::parameterList;
-  using Teuchos::RCP;
-
-  RCP<SolverManagerType> solver = rcp (new SolverManagerType);
-
-  // Some solvers may not like to get a null ParameterList.  If params
-  // is null, replace it with an empty parameter list.  The solver
-  // will fill in default parameters for that case.  Use the name of
-  // the solver's default parameters to name the new empty list.
-  RCP<ParameterList> pl;
-  if (params.is_null()) {
-    pl = parameterList (solver->getValidParameters ()->name ());
-  } else {
-    pl = params;
-  }
-  TEUCHOS_TEST_FOR_EXCEPTION(
-    pl.is_null(), std::logic_error,
-    "Belos::SolverFactory: ParameterList to pass to solver is null.  This "
-    "should never happen.  Please report this bug to the Belos developers.");
-  solver->setParameters (pl);
-  return solver;
+// specialize for Epetra registration
+#ifdef HAVE_BELOS_EPETRA
+template<>
+inline void
+SolverFactoryParent<double, Epetra_MultiVector, Epetra_Operator>::
+registerSolverFactoryForLib () {
+  Belos::Details::Epetra::registerSolverFactory();
 }
+#endif // HAVE_BELOS_EPETRA
+
+// specialize for Tpetra registration - for each of the possible types
+#ifdef HAVE_BELOS_TPETRA
+#define SPECIALIZE_FOR_TPETRA(SC, LO, GO, NT) \
+template<> \
+inline void \
+SolverFactoryParent<SC, ::Tpetra::MultiVector<SC, LO, GO, NT>, \
+  ::Tpetra::Operator<SC, LO, GO, NT>>::registerSolverFactoryForLib () { \
+  Belos::Details::Tpetra::registerSolverFactory(); \
+}
+TPETRA_ETI_MANGLING_TYPEDEFS()
+TPETRA_INSTANTIATE_SLGN_NO_ORDINAL_SCALAR(SPECIALIZE_FOR_TPETRA)
+#endif // HAVE_BELOS_TPETRA
 
-} // namespace Details
+// Fall through case. Register nothing such as for a custom SolverFactory.
+// Then user will be registering their own SolverManagers.
```

Comment:
Is there a "generic registration function," for users to register SolverManagers for their own linear algebra implementation (`Scalar, MV, OP`)?

Reply:
Just realizing now I over thought this - I will change registerCustomSolverFactory() to be named registerGenericSolverFactory(), remove the manual calls, and make it automatically called here. Will update code and this comment once I finish running the tests on this.

Reply:
Added a more general comment on this below. Was remembering now why i didn't just put this into the header due to needing includes. Please see below comment.